src/expr/src/scalar/like_pattern.rs

```diff
@@ -73,3 +158,774 @@ pub fn build_regex(pattern: &str, flags: &str) -> Result<Regex, EvalError> {
         ))),
     }
 }
+
+// The algorithm below is based on the observation that any LIKE pattern can be
+// decomposed into multiple parts:
+//     <PATTERN> := <SUB-PATTERN> (<SUB-PATTERN> ...)
+//     <SUB-PATTERN> := <WILDCARDS> <SUFFIX>
+//
+// The sub-patterns start with zero or more wildcard characters, eventually
+// followed by (non-wildcard) literal characters. The last sub-pattern may
+// have an empty SUFFIX.
+//
+// Example: the PATTERN "n__dl%" can be broken into the following parts:
+//   1. SUB-PATTERN = <WILDCARDS ""> <SUFFIX "n">
+//   2. SUB-PATTERN = <WILDCARDS "__"> <SUFFIX "dl">
+//   3. SUB-PATTERN = <WILDCARDS "%"> <SUFFIX "">
+//
+// The WILDCARDS can be any combination of '_', which matches exactly 1 char,
+// and '%' which matches zero or more chars. These wildcards can be simplified
+// down to the (min, max) of characters they might consume:
+//     ""  = (0, 0)    // doesn't consume any characters
+//     "_" = (1, 1)    // consumes exactly one
+//     "%" = (0, many) // zero or more
+// These are additive, so:
+//     "_%"    = (1, many)
+//     "__%__" = (4, many)
+//     "%%%_"  = (1, many)
+
+// Matches a chain of SUB-PATTERNs
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct StringMatcher {
+    /// The minimum number of characters that can be consumed by the wildcard expression.
+    consume: usize,
+    /// Whether the wildcard expression can consume an arbitrary number of characters.
+    many: bool,
+    /// A string literal that is expected after the wildcards.
+    suffix: String,
+    /// Any sub-patterns that come after this one.
+    next: Option<Box<StringMatcher>>,
```

Comment:
An intrusive linked list strikes me as an odd data structure for this since you don't need constant time insertions or anything. It comes out more idiomatic IMO to use a plain `Vec`, and it's 20% faster on compiling the pattern `a%b_c%d_e%f_g%h_i%j_k%l_` to boot: https://gist.github.com/benesch/66d027bf9674416cd14ee1046e41bfb1

Reply:
Nice, done.